### ISSUE TITLE
How to serialize JSON output at backend?

### ISSUE THREAD

**Issue by vinerz** [ueberdosis/tiptap#241]
Hi there! This question is kinda off-topic, but as you guys are the prosemirror magicians, I'd like to give a try.

I want to save the editor JSON document in my database, but render and filter the XSS at my backend using nodejs. As `prosemirror-model` heavily relies on the browser DOM API, I can't use it there and I also couldn't find any packages at npm capable of doing this out of the box.

Do you guys have any tips for this situation?

Thanks!

**Comment by jjangga0214** [comment]
Another simple strategy might be possible. Just save data as-is without post processing at database. Serialize it When loading that data to frontend. Major frameworks such as vue.js  supports that feature natively.

**Comment by vinerz** [comment]
Hey @jjangga0214, thanks for the input! Unfortunately, I must inject the data back to the frontend as `v-html` so I can't trust user input.

Just for the record, I got it to work by providing a simulated DOM API using JSDOM, if anyone is interested.

It would be very nice to have your opinions on this code as well.

```js
import { JSDOM } from 'jsdom';
import { schema } from 'prosemirror-schema-basic';
import { DOMSerializer } from 'prosemirror-model';

const mainSerializer = DOMSerializer.fromSchema(schema);

const getNewDocument = () => {
  const dom = new JSDOM('<!DOCTYPE html><div id="content"></div>');
  const document = dom.window.document;
  const renderer = document.querySelector('div');

  return { document, renderer };
};

export const prosemirrorToHTML = (nodeObject) => {
  const { document, renderer } = getNewDocument();
  const node = schema.nodeFromJSON(nodeObject);

  mainSerializer.serializeFragment(node, { document }, renderer);
  return document.getElementById('content').innerHTML;
};
```

**Comment by erickwilder** [comment]
@vinerz The the best feature (for me) of Prosemirror is that it is based on a schema that _you have full control_ of. If you get a JSON from the backend (or anywhere else) that doesn't comply with it, the document will not be rendered.

You don't have to use a `v-html` if you store the JSON in the database. Just provide the JSON back when instantiating the editor and Prosemirror will take care of generating the proper markup out of it (a view) in the browser. If something tampers the data, the schema of your editor will not be matched and _only what's defined on it_ is considered as *valid markup*.

**Comment by holtwick** [comment]
For my project [Collect](https://collect-app.com/) I'm also trying to use the JSON document outside of a browser environment. I implemented something similar to #66 to export Markdown. Everything works nice so far, but the problem I run into is, that WebPack bundles TipTap and Vue.js with my code I want to run in my browser less environment (which is [JSContext](https://developer.apple.com/documentation/javascriptcore/jscontext) in my case).

So my proposal would be to completely separate schema related methods from the rest of the TipTap `Node` implementations to allow to generate the schema separated from the UI logic without losing the nice plugin infrastructure. Maybe this could be achieved by subclassing? `NodeSchema` > `NodeLogic`?
